### PR TITLE
Hand-specific PlayerAnimationEvent

### PR DESCRIPTION
**Pull request by Machine-Maker**

### CODE REVIEW

**Comment by SoSeDiK** [comment]
Won't this break plugins that have logic similar to the one below?
```
if (event.getAnimationType() != PlayerAnimationType.ARM_SWING) return; // In case of new animations in future maybe, idk
event.getPlayer().sendMessage("You swung your arm!"); // Event handler logic for when swinging some arm
```
What about PlayerArmSwingEvent which extends PlayerAnimationEvent and adds extra HandEnum/EquipmentSlot/isOffHand() boolean?

**Comment by Machine-Maker** [comment]
Yeah, I thought about that. And it would indeed break that functionality. I am unsure of what the best solution to that would be and an 100% open to suggestions.

Maybe the solution is to add a whole new event in the paper package, I don't know.

File: Spigot-API-Patches/0283-Hand-specific-PlayerAnimationEvent.patch [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
++    /**
++     * Main hand arm swing
++     */
++    MAIN_HAND_ARM_SWING,
++
++    /**
++     * Off hand arm swing
++     */
++    OFFHAND_ARM_SWING,
++    // Paper end
```
Comment (tomcraft) [inline Spigot-API-Patches/0283-Hand-specific-PlayerAnimationEvent.patch]: I think you should keep ARM_SWING and add OFF_ARM_SWING. This way it doesn't break plugins and keep the naming logic with equipment slots. (hand/offhand, arm/offarm)
Cf: https://wiki.vg/index.php?title=Protocol#Entity_Animation_.28clientbound.29

**Comment by kashike** [comment]
Alternatively, keep `ARM_SWING` as the type, but add a `Hand` field/method?

**Comment by tomcraft** [comment]
I just understood this event is triggered by PacketPlayInArmAnimation, i don't know why i was looking at the PacketPlayOutAnimation which has more values and use cases. 
Maybe these ones could be added to the event ?

File: Spigot-API-Patches/0283-Hand-specific-PlayerAnimationEvent.patch (hunk 1 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0283-Hand-specific-PlayerAnimationEvent.patch]: That still results in the same break in the logic SoSeDik described.

**Comment by Machine-Maker** [comment]
> Alternatively, keep `ARM_SWING` as the type, but add a `Hand` field/method?

Yeah, I thought about this as well, but I think the event is too general for that. It’s a PlayerAnimation which could certainly involve something not arm related in the future.

**Comment by Machine-Maker** [comment]
`Rebased after the NMS repackage`

**Comment by Machine-Maker** [comment]
That’s a no from me

**Comment by Machine-Maker** [comment]
Ok, refactored this to add a `PlayerArmSwingEvent` which extends PlayerAnimationEvent. Can get the hand swung from the new event and keeps consistency with existing API.

**Comment by Machine-Maker** [comment]
Rebased for 1.17.1

**Review by Proximyst — APPROVED** [review]

**Review by kennytv — APPROVED** [review]
pufferfish

**Pull request merged**